code/modules/food_and_drinks/pizzabox.dm

```diff
@@ -321,37 +334,50 @@
 	var/static/list/pizza_preferences
+	///Who the box is currently attuned to. Prevents it re-attuning needlessly.
+	var/mob/living/carbon/human/attuned_to
 
 /obj/item/pizzabox/infinite/Initialize()
 	. = ..()
 	if(!pizza_preferences)
 		pizza_preferences = list()
 
 /obj/item/pizzabox/infinite/examine(mob/user)
+	if(!open && ishuman(user))
+		attune_pizza(user) //pizza tag changes based on examiner
 	. = ..()
 	if(isobserver(user))
 		. += "<span class='deadsay'>This pizza box is anomalous, and will produce infinite pizza.</span>"
 
 /obj/item/pizzabox/infinite/attack_self(mob/living/user)
-	QDEL_NULL(pizza)
 	if(ishuman(user))
 		attune_pizza(user)
-	. = ..()
+	return ..()
 
-/obj/item/pizzabox/infinite/proc/attune_pizza(mob/living/carbon/human/noms) //tonight on "proc names I never thought I'd type"
-	if(!pizza_preferences[noms.ckey])
-		pizza_preferences[noms.ckey] = pickweight(pizza_types)
-		if(noms.has_quirk(/datum/quirk/pineapple_liker))
-			pizza_preferences[noms.ckey] = /obj/item/food/pizza/pineapple
-		else if(noms.has_quirk(/datum/quirk/pineapple_hater))
-			var/list/pineapple_pizza_liker = pizza_types.Copy()
-			pineapple_pizza_liker -= /obj/item/food/pizza/pineapple
-			pizza_preferences[noms.ckey] = pickweight(pineapple_pizza_liker)
-		else if(noms.mind && noms.mind.assigned_role == "Botanist")
-			pizza_preferences[noms.ckey] = /obj/item/food/pizza/dank
-
-	var/obj/item/pizza_type = pizza_preferences[noms.ckey]
-	pizza = new pizza_type (src)
-	pizza.foodtypes = noms.dna.species.liked_food //it's our favorite!
+/obj/item/pizzabox/infinite/proc/attune_pizza(mob/living/carbon/human/nommer) //tonight on "proc names I never thought I'd type"
+	if(nommer.ckey)
+		//list our ckey and assign a favourite pizza
+		if(!pizza_preferences[nommer.ckey])
+			pizza_preferences[nommer.ckey] = pickweight(pizza_types)
+			if(nommer.has_quirk(/datum/quirk/pineapple_liker))
+				pizza_preferences[nommer.ckey] = /obj/item/food/pizza/pineapple
+			else if(nommer.has_quirk(/datum/quirk/pineapple_hater))
+				var/list/pineapple_pizza_liker = pizza_types.Copy()
+				pineapple_pizza_liker -= /obj/item/food/pizza/pineapple
+				pizza_preferences[nommer.ckey] = pickweight(pineapple_pizza_liker)
+			else if(nommer.mind && nommer.mind.assigned_role == "Botanist")
+				pizza_preferences[nommer.ckey] = /obj/item/food/pizza/dank
+		//delete the current pizza if it isn't for us
+		if(pizza && attuned_to && attuned_to != nommer)
+			QDEL_NULL(pizza)
+		//if there's no pizza and we have a favourite, create it and update the boxtag
+		if(!pizza && pizza_preferences[nommer.ckey])
+			var/obj/item/food/pizza/favourite_pizza_type = pizza_preferences[nommer.ckey]
+			pizza = new favourite_pizza_type
+			boxtag_set = FALSE
+			update_icon() //update our boxtag to match our new pizza
+			pizza.foodtypes = nommer.dna.species.liked_food //it's our favorite!
+			attuned_to = nommer
```

Comment:
What happens if the `nommer` gets qdeleted before the pizza?
Will this variable point to a qdeleted entity, forcing a hard deletion?
Does this variable get cleaned anywhere, such as on the object's `Destroy()`?

If you don't understand how reference handling and hard deletions work, poke me on the coder channel and I'll explain.

Reply:
So I wasn't sure how to clear attuned_to in the event the eater is destroyed (as an aside it seems misguided to me to actually delete a player mob) so I just made it "respawn" the pizza in the box if the type is not the human's favourite or the foodtype isn't what the human likes.

When I tested this with an ash walker and a human, the human didn't get any adverse taste messages but the ash walker got them on both its own pizza and the human's pizza. I don't think I introduced that, though. If not, I may not try to fix it.